File: apps/web/src/features/billing/service.ts
import { apiRequest, ApiError, pageSchema, type Page } from "@/services/api-client";
import { demoDelay, select } from "@/lib/data-mode";
import { demoBusiness, mulMoney, nextNumber, sumMoney, taxOf } from "@/demo/business";
import { demoId, matches, paginate } from "@/demo/store";
import { dateOnly } from "@/demo/random";
import { toCents, centsToString } from "@/lib/format";
import {
  billingSummarySchema,
  invoiceDetailSchema,
  invoiceSchema,
  paymentSchema,
  type BillingSummary,
  type Invoice,
  type InvoiceDetail,
  type InvoiceInput,
  type ListParams,
  type Payment,
  type PaymentInput,
} from "@/features/business/types";

export interface BillingService {
  summary(): Promise<BillingSummary>;
  invoices(params: ListParams & { customerId?: string; overdue?: boolean }): Promise<Page<Invoice>>;
  invoice(id: string): Promise<InvoiceDetail>;
  createInvoice(input: InvoiceInput): Promise<InvoiceDetail>;
  invoiceAction(id: string, action: "issue" | "void"): Promise<InvoiceDetail>;
  recordPayment(invoiceId: string, input: PaymentInput): Promise<Payment>;
  payments(params: { page?: number; pageSize?: number }): Promise<Page<Payment & { invoice_number?: string; customer_name?: string | null }>>;
}

const live: BillingService = {
  summary: () => apiRequest("GET", "/billing/summary", billingSummarySchema),
  invoices: ({ page = 1, pageSize = 25, search, status, customerId, overdue }) =>
    apiRequest("GET", "/billing/invoices", pageSchema(invoiceSchema), {
      query: { page, page_size: pageSize, search, status, customer_id: customerId, overdue: overdue || undefined },
    }),
  invoice: (id) => apiRequest("GET", `/billing/invoices/${id}`, invoiceDetailSchema),
  createInvoice: (input) => apiRequest("POST", "/billing/invoices", invoiceDetailSchema, { body: input }),
  invoiceAction: (id, action) => apiRequest("POST", `/billing/invoices/${id}/actions`, invoiceDetailSchema, { body: { action } }),
  recordPayment: (id, input) => apiRequest("POST", `/billing/invoices/${id}/payments`, paymentSchema, { body: input }),
  payments: ({ page = 1, pageSize = 25 }) =>
    apiRequest("GET", "/billing/payments", pageSchema(paymentSchema), { query: { page, page_size: pageSize } }),
};

function rule(code: string, message: string): never {
  throw new ApiError(422, code, undefined, message);
}

function view(invoice: InvoiceDetail): InvoiceDetail {
  const overdue = ["issued", "partially_paid"].includes(invoice.status) && !!invoice.due_date && invoice.due_date < dateOnly(0);
  const actions =
    invoice.status === "draft" ? ["issue", "void"] :
    invoice.status === "issued" ? (toCents(invoice.amount_paid) > BigInt(0) ? ["record_payment"] : ["record_payment", "void"]) :
    invoice.status === "partially_paid" ? ["record_payment"] : [];
  return { ...invoice, is_overdue: overdue, balance_due: sumMoney([invoice.total, `-${invoice.amount_paid}`]), next_actions: actions, lines: [...invoice.lines], payments: [...invoice.payments] };
}

function find(id: string) {
  const invoice = demoBusiness().invoices.find((i) => i.id === id);
  if (!invoice) throw new ApiError(404, "RESOURCE_NOT_FOUND");
  return invoice;
}

const demo: BillingService = {
  async summary() {
    await demoDelay();
    const business = demoBusiness();
    const open = business.invoices.filter((i) => ["issued", "partially_paid"].includes(i.status)).map(view);
    const overdue = open.filter((i) => i.is_overdue);
    const monthStart = dateOnly(0).slice(0, 8) + "01";
    const collected = business.invoices.flatMap((i) => i.payments).filter((p) => p.received_on >= monthStart);
    return {
      currency: business.settings.default_currency,
      outstanding: sumMoney(open.map((i) => i.balance_due)),
      overdue: sumMoney(overdue.map((i) => i.balance_due)),
      overdue_count: overdue.length,
      collected_this_month: sumMoney(collected.map((p) => p.amount)),
      draft_count: business.invoices.filter((i) => i.status === "draft").length,
    };
  },
  async invoices({ page = 1, pageSize = 25, search, status, customerId, overdue }) {
    await demoDelay();
    const rows = demoBusiness()
      .invoices.map(view)
      .filter((i) => (!status || i.status === status) && (!customerId || i.customer_id === customerId) && (!overdue || i.is_overdue) && (!search || matches(i.number, search) || matches(i.customer_name, search)))
      .sort((a, b) => b.created_at.localeCompare(a.created_at));
    return paginate(rows.map(({ lines: _l, payments: _p, next_actions: _n, ...i }) => i), page, pageSize);
  },
  async invoice(id) {
    await demoDelay();
    return view(find(id));
  },
  async createInvoice(input) {
    await demoDelay(350);
    const business = demoBusiness();
    const customer = business.customers.find((c) => c.id === input.customer_id);
    if (!customer) throw new ApiError(404, "RESOURCE_NOT_FOUND");
    const lines = input.lines.map((l, i) => {
      const gross = mulMoney(l.unit_price, l.quantity);
      if (toCents(l.discount) > toCents(gross)) rule("INVALID_DISCOUNT", "A discount exceeds its line");
      return { id: demoId("line"), variant_id: null, position: i + 1, description: l.description, quantity: l.quantity, unit_price: l.unit_price, discount: l.discount, line_total: sumMoney([gross, `-${l.discount}`]), gross };
    });
    const subtotal = sumMoney(lines.map((l) => l.gross));
    const discount_total = sumMoney(lines.map((l) => l.discount));
    const taxable = sumMoney([subtotal, `-${discount_total}`]);
    const tax_total = taxOf(taxable, business.settings.tax_rate);
    const total = sumMoney([taxable, tax_total]);
    const invoice: InvoiceDetail = {
      id: demoId("inv"), number: nextNumber(business, "invoice"), customer_id: customer.id, order_id: null, status: "draft", issue_date: null,
      due_date: input.due_date ?? null, currency: business.settings.default_currency, subtotal, discount_total, tax_total, total, amount_paid: "0.00",
      notes: input.notes, created_at: new Date().toISOString(), customer_name: customer.name, balance_due: total, is_overdue: false,
      lines: lines.map(({ gross: _g, ...l }) => l), payments: [], next_actions: [],
    };
    business.invoices.unshift(invoice);
    return view(invoice);
  },
  async invoiceAction(id, action) {
    await demoDelay(250);
    const invoice = find(id);
    if (action === "issue") {
      if (invoice.status !== "draft") rule("INVALID_TRANSITION", "Only drafts can be issued");
      invoice.status = "issued";
      invoice.issue_date = dateOnly(0);
      invoice.due_date ??= dateOnly(demoBusiness().settings.invoice_due_days);
    } else {
      if (toCents(invoice.amount_paid) > BigInt(0)) rule("INVOICE_HAS_PAYMENTS", "Paid invoices cannot be voided");
      if (!["draft", "issued"].includes(invoice.status)) rule("INVALID_TRANSITION", `An invoice cannot move from ${invoice.status} to void`);
      invoice.status = "void";
    }
    return view(invoice);
  },
  async recordPayment(invoiceId, input) {
    await demoDelay(300);
    const business = demoBusiness();
    const invoice = find(invoiceId);
    if (!["issued", "partially_paid"].includes(invoice.status)) rule("INVOICE_NOT_OPEN", "Payments need an issued invoice");
    const balance = toCents(invoice.total) - toCents(invoice.amount_paid);
    const amount = toCents(input.amount);
    if (amount <= BigInt(0)) rule("INVALID_AMOUNT", "Enter an amount greater than zero");
    if (amount > balance) rule("OVERPAYMENT", "Payment exceeds the balance due");
    const payment: Payment = {
      id: demoId("pay"), invoice_id: invoice.id, number: nextNumber(business, "payment"), amount: centsToString(amount), currency: invoice.currency,
      method: input.method, received_on: input.received_on || dateOnly(0), reference: input.reference, recorded_by_label: "Amina Rahman", created_at: new Date().toISOString(),
    };
    invoice.payments.push(payment);
    invoice.amount_paid = centsToString(toCents(invoice.amount_paid) + amount);
    invoice.status = toCents(invoice.amount_paid) === toCents(invoice.total) ? "paid" : "partially_paid";
    return payment;
  },
  async payments({ page = 1, pageSize = 25 }) {
    await demoDelay();
    const rows = demoBusiness()
      .invoices.flatMap((i) => i.payments.map((p) => ({ ...p, invoice_number: i.number, customer_name: i.customer_name })))
      .sort((a, b) => b.received_on.localeCompare(a.received_on));
    return paginate(rows, page, pageSize);
  },
};

export const billingService = select<BillingService>({ demo, live });

File: apps/api/app/modules/billing/service.py
from collections.abc import Sequence
from datetime import UTC, date, datetime, timedelta
from decimal import Decimal
from uuid import UUID

from sqlalchemy import func, select
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.pagination import Page, Pagination
from app.integrations.outbox import EntityRef, emit
from app.modules.audit.service import record
from app.modules.billing.models import Invoice, InvoiceLine, Payment
from app.modules.billing.schemas import (
    BillingSummary,
    InvoiceCreate,
    InvoiceDetail,
    InvoiceLineView,
    InvoiceListItem,
    InvoiceView,
    PaymentCreate,
    PaymentView,
)
from app.modules.business_settings.service import get_settings_row
from app.modules.customers.models import Customer
from app.modules.customers.service import log_activity
from app.modules.orders.models import Order, OrderLine
from app.shared.errors import BusinessRuleViolation
from app.shared.money import compute_totals, quantize
from app.shared.scope import WorkspaceScope
from app.shared.sequences import next_number
from app.shared.state_machine import StateMachine
from app.shared.workspace_repository import WorkspaceRepository, like_pattern, to_page

INVOICE_STATES = StateMachine(
    "invoice",
    {
        "draft": frozenset({"issued", "void"}),
        "issued": frozenset({"partially_paid", "paid", "void"}),
        "partially_paid": frozenset({"paid"}),
        "paid": frozenset(),
        "void": frozenset(),
    },
)
OPEN_STATUSES = ("issued", "partially_paid")


def balance(invoice: Invoice) -> Decimal:
    return quantize(invoice.total - invoice.amount_paid)


def _invoice_payload(invoice: Invoice) -> dict[str, str | None]:
    return {
        "invoice_id": str(invoice.id),
        "number": invoice.number,
        "status": invoice.status,
        "customer_id": str(invoice.customer_id) if invoice.customer_id else None,
        "order_id": str(invoice.order_id) if invoice.order_id else None,
        "total": str(invoice.total),
        "amount_paid": str(invoice.amount_paid),
        "currency": invoice.currency,
    }


class BillingService:
    def __init__(self, session: AsyncSession, scope: WorkspaceScope) -> None:
        self.session, self.scope = session, scope
        self.invoices = WorkspaceRepository(session, Invoice, scope)
        self.lines = WorkspaceRepository(session, InvoiceLine, scope)
        self.payments = WorkspaceRepository(session, Payment, scope)
        self.customers = WorkspaceRepository(session, Customer, scope)

    def _item(self, invoice: Invoice, customer_name: str | None) -> InvoiceListItem:
        overdue = (
            invoice.status in OPEN_STATUSES
            and invoice.due_date is not None
            and invoice.due_date < date.today()
        )
        return InvoiceListItem(
            **InvoiceView.model_validate(invoice).model_dump(),
            customer_name=customer_name,
            balance_due=balance(invoice),
            is_overdue=overdue,
        )

    async def _names(self, ids: set[UUID]) -> dict[UUID, str]:
        if not ids:
            return {}
        rows = await self.session.execute(
            select(Customer.id, Customer.name).where(
                self.customers.predicate(), Customer.id.in_(ids)
            )
        )
        return {i: n for i, n in rows}

    async def search(
        self,
        page: Pagination,
        status: str | None = None,
        customer_id: UUID | None = None,
        overdue: bool = False,
        search: str | None = None,
    ) -> Page[InvoiceListItem]:
        statement = self.invoices.select()
        if status:
            statement = statement.where(Invoice.status == status)
        if customer_id:
            statement = statement.where(Invoice.customer_id == customer_id)
        if overdue:
            statement = statement.where(
                Invoice.status.in_(OPEN_STATUSES), Invoice.due_date < date.today()
            )
        if search:
            statement = statement.where(Invoice.number.ilike(like_pattern(search.strip())))
        statement = statement.order_by(Invoice.created_at.desc(), Invoice.id)
        rows, total = await self.invoices.page(statement, page)
        names = await self._names({r.customer_id for r in rows})
        items = [self._item(r, names.get(r.customer_id)) for r in rows]
        return Page(items=items, total=total, page=page.page, page_size=page.page_size)

    async def detail(self, invoice_id: UUID) -> InvoiceDetail:
        invoice = await self.invoices.get(invoice_id)
        lines = await self.session.scalars(
            self.lines.select()
            .where(InvoiceLine.invoice_id == invoice.id)
            .order_by(InvoiceLine.position)
        )
        payments = await self.session.scalars(
            self.payments.select()
            .where(Payment.invoice_id == invoice.id)
            .order_by(Payment.received_on, Payment.created_at)
        )
        names = await self._names({invoice.customer_id})
        actions = {
            "draft": ["issue", "void"],
            "issued": ["record_payment", "void"]
            if invoice.amount_paid == 0
            else ["record_payment"],
            "partially_paid": ["record_payment"],
        }.get(invoice.status, [])
        return InvoiceDetail(
            **self._item(invoice, names.get(invoice.customer_id)).model_dump(),
            lines=[InvoiceLineView.model_validate(x) for x in lines],
            payments=[PaymentView.model_validate(p) for p in payments],
            next_actions=actions,
        )

    async def create(self, data: InvoiceCreate) -> Invoice:
        self.scope.require("billing.write")
        await self.customers.get(data.customer_id)
        settings = await get_settings_row(self.session, self.scope)
        try:
            totals_by_line, totals = compute_totals(
                [(x.quantity, x.unit_price, x.discount) for x in data.lines], settings.tax_rate
            )
        except ValueError:
            raise BusinessRuleViolation("INVALID_DISCOUNT", "A discount exceeds its line") from None
        invoice = await self.invoices.add(
            self.invoices.new(
                number=await next_number(self.session, self.scope, "invoice"),
                customer_id=data.customer_id,
                currency=settings.default_currency,
                due_date=data.due_date,
                notes=data.notes,
                subtotal=totals.subtotal,
                discount_total=totals.discount_total,
                tax_total=totals.tax_total,
                total=totals.total,
            )
        )
        for position, (line, line_total) in enumerate(
            zip(data.lines, totals_by_line, strict=True), start=1
        ):
            self.session.add(
                self.lines.new(
                    invoice_id=invoice.id,
                    position=position,
                    description=line.description,
                    quantity=line.quantity,
                    unit_price=line.unit_price,
                    discount=line.discount,
                    line_total=line_total,
                )
            )
        await self.session.flush()
        await record(
            self.session,
            "invoice.created",
            scope=self.scope,
            entity_type="invoice",
            entity_id=invoice.id,
            details={"number": invoice.number, "total": str(invoice.total)},
        )
        return invoice

    async def create_from_order(self, order: Order, lines: Sequence[OrderLine]) -> Invoice:
        """Issue an invoice for a confirmed order inside the order's transaction."""
        self.scope.require("billing.write")
        existing = await self.invoices.find(Invoice.order_id == order.id, Invoice.status != "void")
        if existing is not None:
            return existing
        settings = await get_settings_row(self.session, self.scope)
        today = date.today()
        invoice = await self.invoices.add(
            self.invoices.new(
                number=await next_number(self.session, self.scope, "invoice"),
                customer_id=order.customer_id,
                order_id=order.id,
                status="issued",
                issue_date=today,
                due_date=today + timedelta(days=settings.invoice_due_days),
                currency=order.currency,
                subtotal=order.subtotal,
                discount_total=order.discount_total,
                tax_total=order.tax_total,
                total=order.total,
            )
        )
        for line in lines:
            self.session.add(
                self.lines.new(
                    invoice_id=invoice.id,
                    variant_id=line.variant_id,
                    position=line.position,
                    description=line.description,
                    quantity=Decimal(line.quantity),
                    unit_price=line.unit_price,
                    discount=line.discount,
                    line_total=line.line_total,
                )
            )
        await self.session.flush()
        await log_activity(
            self.session,
            self.scope,
            order.customer_id,
            "invoice",
            f"Invoice {invoice.number} issued for {order.number}",
            "invoice",
            invoice.id,
        )
        await record(
            self.session,
            "invoice.issued_from_order",
            scope=self.scope,
            entity_type="invoice",
            entity_id=invoice.id,
            details={"order": order.number, "total": str(invoice.total)},
        )
        await emit(
            self.session,
            self.scope,
            "invoice.issued",
            _invoice_payload(invoice),
            EntityRef("invoice", invoice.id),
        )
        return invoice

    async def act(self, invoice_id: UUID, action: str) -> Invoice:
        self.scope.require("billing.write")
        invoice = await self.invoices.get(invoice_id, for_update=True)
        if action == "issue":
            INVOICE_STATES.ensure(invoice.status, "issued")
            invoice.status = "issued"
            invoice.issue_date = date.today()
            if invoice.due_date is None:
                settings = await get_settings_row(self.session, self.scope)
                invoice.due_date = date.today() + timedelta(days=settings.invoice_due_days)
        elif action == "void":
            if invoice.amount_paid > 0:
                raise BusinessRuleViolation(
                    "INVOICE_HAS_PAYMENTS", "Paid invoices cannot be voided"
                )
            INVOICE_STATES.ensure(invoice.status, "void")
            invoice.status, invoice.voided_at = "void", datetime.now(UTC)
        else:
            raise BusinessRuleViolation("UNKNOWN_ACTION", "Unknown invoice action")
        await self.session.flush()
        await record(
            self.session,
            f"invoice.{action}",
            scope=self.scope,
            entity_type="invoice",
            entity_id=invoice.id,
        )
        if action == "issue":
            await emit(
                self.session,
                self.scope,
                "invoice.issued",
                _invoice_payload(invoice),
                EntityRef("invoice", invoice.id),
            )
        return invoice

    async def void_for_order(self, order_id: UUID) -> None:
        invoice = await self.invoices.find(Invoice.order_id == order_id, Invoice.status != "void")
        if invoice is None:
            return
        await self.act(invoice.id, "void")

    async def record_payment(self, invoice_id: UUID, data: PaymentCreate) -> Payment:
        self.scope.require("billing.write")
        invoice = await self.invoices.get(invoice_id, for_update=True)
        if invoice.status not in OPEN_STATUSES:
            raise BusinessRuleViolation("INVOICE_NOT_OPEN", "Payments need an issued invoice")
        amount = quantize(data.amount)
        if amount > balance(invoice):
            raise BusinessRuleViolation("OVERPAYMENT", "Payment exceeds the balance due")
        payment = await self.payments.add(
            self.payments.new(
                invoice_id=invoice.id,
                number=await next_number(self.session, self.scope, "payment"),
                amount=amount,
                currency=invoice.currency,
                method=data.method,
                received_on=data.received_on or date.today(),
                reference=data.reference,
                recorded_by_label=self.scope.actor_label[:80],
            )
        )
        invoice.amount_paid = quantize(invoice.amount_paid + amount)
        target = "paid" if invoice.amount_paid == invoice.total else "partially_paid"
        if target != invoice.status:
            INVOICE_STATES.ensure(invoice.status, target)
            invoice.status = target
        await self.session.flush()
        await log_activity(
            self.session,
            self.scope,
            invoice.customer_id,
            "payment",
            f"Payment {payment.number} of {invoice.currency} {amount} on {invoice.number}",
            "payment",
            payment.id,
        )
        await record(
            self.session,
            "payment.recorded",
            scope=self.scope,
            entity_type="payment",
            entity_id=payment.id,
            details={"invoice": invoice.number, "amount": str(amount)},
        )
        await emit(
            self.session,
            self.scope,
            "payment.received",
            {
                "payment_id": str(payment.id),
                "number": payment.number,
                "invoice_id": str(invoice.id),
                "amount": str(amount),
                "currency": invoice.currency,
                "method": payment.method,
            },
            EntityRef("payment", payment.id),
        )
        if invoice.status == "paid":
            await emit(
                self.session,
                self.scope,
                "invoice.paid",
                _invoice_payload(invoice),
                EntityRef("invoice", invoice.id),
            )
        return payment

    async def payments_page(self, page: Pagination) -> Page[PaymentView]:
        statement = self.payments.select().order_by(Payment.received_on.desc(), Payment.id)
        rows, total = await self.payments.page(statement, page)
        return to_page(PaymentView, rows, total, page)

    async def customer_balance(self, customer_id: UUID) -> Decimal:
        value = await self.session.scalar(
            select(func.coalesce(func.sum(Invoice.total - Invoice.amount_paid), 0)).where(
                self.invoices.predicate(),
                Invoice.customer_id == customer_id,
                Invoice.status.in_(OPEN_STATUSES),
            )
        )
        return quantize(Decimal(value or 0))

    async def summary(self) -> BillingSummary:
        settings = await get_settings_row(self.session, self.scope)
        open_balance = Invoice.total - Invoice.amount_paid
        outstanding = await self.session.scalar(
            select(func.coalesce(func.sum(open_balance), 0)).where(
                self.invoices.predicate(),
                Invoice.status.in_(OPEN_STATUSES),
                Invoice.currency == settings.default_currency,
            )
        )
        overdue_row = (
            await self.session.execute(
                select(func.coalesce(func.sum(open_balance), 0), func.count()).where(
                    self.invoices.predicate(),
                    Invoice.status.in_(OPEN_STATUSES),
                    Invoice.due_date < date.today(),
                    Invoice.currency == settings.default_currency,
                )
            )
        ).one()
        month_start = date.today().replace(day=1)
        collected = await self.session.scalar(
            select(func.coalesce(func.sum(Payment.amount), 0)).where(
                self.payments.predicate(),
                Payment.received_on >= month_start,
                Payment.currency == settings.default_currency,
            )
        )
        drafts = await self.session.scalar(
            select(func.count())
            .select_from(Invoice)
            .where(self.invoices.predicate(), Invoice.status == "draft")
        )
        return BillingSummary(
            currency=settings.default_currency,
            outstanding=quantize(Decimal(outstanding or 0)),
            overdue=quantize(Decimal(overdue_row[0] or 0)),
            overdue_count=int(overdue_row[1] or 0),
            collected_this_month=quantize(Decimal(collected or 0)),
            draft_count=int(drafts or 0),
        )
